public/app/plugins/datasource/loki/queryUtils.ts

```diff
@@ -295,3 +318,105 @@ export function getStreamSelectorsFromQuery(query: string): string[] {
 
   return labelMatchers;
 }
+
+const PARTITION_LIMIT = 60;
+export function partitionTimeRange(range: TimeRange, timeShift = 1, unit: DurationUnit = 'm'): TimeRange[] {
+  const partition: TimeRange[] = [];
+  for (let from = dateTime(range.from), to = dateTime(from).add(timeShift, unit); to <= range.to && from < range.to; ) {
+    partition.push({
+      from,
+      to,
+      raw: { from, to },
+    });
+
+    /**
+     * The user can request for any arbitrary time range. If we receive one from too
+     * long ago, this will cause an extremely large loop which could break the app.
+     * Additionally, data retention is limited, so ranges that go above PARTIION_LIMIT
+     * will be ignored.
+     */
+    if (partition.length > PARTITION_LIMIT) {
+      return [range];
+    }
+
+    from = dateTime(to);
+    to = dateTime(to).add(timeShift, unit);
+
+    if (to > range.to) {
+      to = range.to;
+    }
+  }
+
+  return partition;
+}
+
+export function requestSupportsPartitioning(queries: LokiQuery[]) {
+  /*
+   * For now, we would not split when more than 1 query is requested.
+   */
+  if (queries.length > 1) {
+    return false;
+  }
+
+  if (isLogsQuery(queries[0].expr)) {
+    return false;
+  }
+
+  if (isRateQuery(queries[0].expr)) {
+    return false;
+  }
+
+  return true;
+}
+
+export function combineResponses(currentResult: DataQueryResponse | null, newResult: DataQueryResponse) {
+  if (!currentResult) {
+    return newResult;
+  }
+
+  newResult.data.forEach((newFrame) => {
+    const currentFrame = currentResult.data.find((frame) => frame.name === newFrame.name);
+    const currentIndex = currentResult.data.findIndex((frame) => frame.name === newFrame.name);
+
+    if (!currentFrame) {
+      currentResult.data.push(newFrame);
+      return;
+    }
+    const frame = new MutableDataFrame(currentFrame);
+    combineFrames(frame, newFrame);
+    currentResult.data[currentIndex] = frame;
+  });
+
+  return currentResult;
+}
+
+function combineFrames(dest: MutableDataFrame, source: MutableDataFrame) {
+  dest.reverse();
+  for (let j = source.fields[0].values.length - 1; j > 0; j--) {
+    dest.fields[0].values.add(source.fields[0].values.get(j));
+    dest.fields[1].values.add(source.fields[1].values.get(j));
+  }
+  dest.reverse();
+}
+
+/**
+ * Checks if the current response has reached the requested amount of results or not.
+ * For log queries, we will ensure that the current amount of results doesn't go beyond `maxLines`.
+ */
+export function resultLimitReached(request: DataQueryRequest<LokiQuery>, result: DataQueryResponse) {
```

Comment:
If we enable it in `requestSupportsPartitioning()`, this function will ensure that we respect the requested line limit of logs.

Reply:
nice 👍 . i recommend to implement this logs-functionality in a separate pull request though, that way we can focus on that problem separately.